frontend/model/contracts/group.js

```diff
@@ -256,6 +256,33 @@ DefineContract({
         archiveProposal(state, data.proposalHash)
       }
     },
+    'gi.contracts/group/removeMember': {
```

Comment:
@taoeffect so, some updates about this:

1. I figured out how to remove contracts. After all, `removeContract` already existed! I think that's what we need right?
2. The member is removed as expected (I think!). It's only missing to redirect them to home or `/dashboard`. How can we do that with sbp?
3. Now, a very weird thing happens. If the removed member joins the group again, it doesn't work! Because when syncing the contract, one of the actions is to remove them... so it kinda does the opposite. Any suggestions on this?
4. The proposal layout is also done, it's only missing the action [VOTE_FOR] when the proposal gets accepted, but to do that we need to first address the previous points.

Another thing:
- When a member is removed, there's an error on the remaining members' console: `uncaught Vue error in render: TypeError: "_vm.member is undefined"` Couldn't find yet what's causing this, but working on it.

Reply:
### Recap

To remove a member, we basically need to undo everything that happens when they join, i.e. everything in `'gi.contracts/group/inviteAccept'`.

`'gi.contracts/group/inviteAccept'` does two things (that I know you noticed, based on this PR):

1. In the `/process` function, it adds their group profile to `state.profiles` (and updates the invite, although the invite is irrelevant for our purposes here)
2. In the `sideEffect`, we call `state/enqueueContractSync` on their identity contract, so that we have access to their global profile, which includes things like their `displayName` and their profile picture URL

In `proposals.js`, we see a TODO comment:

```js
    [VOTE_FOR]: function (state, { proposalHash, passPayload }) {
      console.error('unimplemented!')
      // TODO: unsubscribe from their mailbox and identity contract
      //       call commit('removeContract'), etc.
    },
```

Note that this comment is a guide only, as it contains one mistake: we never subscribe to anyone's mailbox contract other than our own, so we don't need to unsubscribe from that.

We are subscribed to their identity contract, however, as well as our own identity contract. We are subscribed to theirs any time we interact with the user in any way, including outside of a group, and we are subscribed to our own identity contract always (unless we are logged out). So if a group kicks us out, we do not need to unsubscribe from our identity contract (or our mailbox contract). But if someone else is kicked out, we *might* need to unsubscribe from their identity contract — **IF we have no other connection to this person.**

Imagine if a friend of ours gets kicked out of a group, but we have an open DM chat with them. We need to be subscribed to their identity contract still, if the chat is outside of the group. Otherwise we won't be able to see their profile picture or their `displayName`. Basically, if we're interacting with someone at all in any way, we need to be subscribed to their identity contract.

So for now, as a temporary solution, let's *not* unsubscribe from their identity contract if they're kicked out of the group, and open a separate issue to deal with keeping track of when we should unsubscribe from a user's identity contract.

### Also, some points of note:

- Subscribing and unsubscribing from contracts is, as you've observed correctly in this PR, something that must be done in a `sideEffect`, because it is impacting state that is outside of this current contract. Only state related to this contract directly can be modified in the `/process` function.

### Answering your questions

> 1. I figured out how to remove contracts. After all, `removeContract` already existed! I think that's what we need right?

Yes, we can use that to stop listening for events on the group contract if we are the ones being removed.

> 2. The member is removed as expected (I think!). It's only missing to redirect them to home or `/dashboard`. How can we do that with sbp?

The dashboard is only something visible in the group. We can redirect them home, sure. That is something that can be done via `sbp` if it is an action that is made available through `sbp`. Since we are using Vue, we can tell the root vue router to navigate us to `/` in the `sideEffect`.

> 3. Now, a very weird thing happens. If the removed member joins the group again, it doesn't work! Because when syncing the contract, one of the actions is to remove them... so it kinda reverts the whole thing. Any suggestions to overcome this?

That's a funny bug. Some observations:

- It is OK if during the `'gi.contracts/group/proposalVote/process'` the user's profile is removed from `state.profiles`, because it should be added back again in a future message (once we get to `'gi.contracts/group/inviteAccept/process'`).
- So the problem is likely in the fact that the `sideEffect` is asking us to unsubscribe from the group that we're currently in the middle of syncing

The issue is, if we're offline when someone proposes to remove us from a group, and the proposal passes, then we do in fact want to unsubscribe ourselves from the group once `enqueueContractSync` gets called after we login to GI.

In other words, sometimes we want to ignore it, and sometimes we don't. We want to ignore it only when we are in the middle of a contract sync that is a direct result of joining the group. 

So this suggests a few things:

1. We need to look into the code that's in Join.vue. In that code, we need to set a global flag that says we're calling `enqueueContractSync` because we're just joining a group, and unset that flag after the sync is finished.
2. This updated code should be moved out of Join.vue and into `controller/actions/group.js`.
3. Then in the `sideEffect` function for `'gi.contracts/group/proposalVote'`, if we see that a vote has passed removing us from this same group, we need to NOT remove ourselves if that flag is currently set

For example, I had a similar problem to deal with when implementing `autoBanSenderOfMessage` in `state.js`, and I did this:

```js
      if (contractIsSyncing[contractID]) {
        console.warn(`skipping autoBanSenderOfMessage since we're syncing ${contractID}`)
        return // don't do anything, assume the existing users handled this event
      }
```

Your solution should be a bit cleaner than that. :-)

> 4. The proposal layout is also done, it's only missing the action [VOTE_FOR] when the proposal gets accepted, but to do that we need to first address the previous points.

Yes, also, per our convo on Slack, this contract action `'gi.contracts/group/removeMember'` should be removed, since we want this code tot be done only in the `[VOTE_FOR]` (so that it's only run as a result of a successful proposal). The sideEffect can be moved into the sideEffect for `'gi.contracts/group/proposalVote'`

Reply:
Great overview, it explained a lot and was enough to help me finish the member removal! Here's some notes about it:
- I used `sbp('okTurtles.data/set', 'JOINING_GROUP', true|false)` to fix the _funny_ bug when rejoining a group.
- I created #828 (and pointed it on the code) to later address the condition to remove an identity contract.
- I manually tested the following scenarios where user1 and user2 have a group in common:
  - ✅ User1 and user2 are both online. user1 removes user2. user2 is redirected to Home.
  - ✅ user1 removes user2 while user2 is offline. user2 logs in and sees screen to create a new group.
  - ✅ User1 and user2 are both online. user2 also has a 2nd group. user1 removes user2. user2 is redirected to the dashboard of their 2nd group. 
  - ✅ User1 and user2. user2 also has a 2nd group. user1 removes user2 while user2 is offline. user2 logs in and sees the dashboard of its 2nd group. 
  - any other cases we need to cover? I'll write the cypress tests later for the scenarios where user2 is offline during the removal.

You said this:
 > contract action 'gi.contracts/group/removeMember' should be removed, since we want this code to be done only in the [VOTE_FOR] (so that it's only run as a result of a successful proposal). The sideEffect can be moved into the sideEffect for 'gi.contracts/group/proposalVote'

Okay, I can move it to there... but then, when there are only 2 members in group, how do we reach that code, taking in account this isn't a proposal?

Reply:
Those tests sounds great to me!

> Okay, I can move it to there... but then, when there are only 2 members in group, how do we reach that code, taking in account this isn't a proposal?

Right now we don't. All I am saying is that this code needs to be moved for now. We can create an issue to address this in greater detail later, when we add a permissions and authorizations system. I.e. it would make sense to have this code here *if* we had such a system, where someone, e.g. an `admin`, can unilaterally send this message without a vote, and then in this code we would have a check to see if the user who sent this message has the `admin` permission.

But let's leave that for an issue for later, and for now move this into the proposal.js section of `[VOTE_FOR]`

Reply:
So, @taoeffect what you are saying is that for now, we cannot remove a member, unless it's a proposal. Is that right?

Reply:
Yes exactly, for now (while we haven't implemented permissions), removal should only be done through proposals. This is something we can rectify later, once permissions are implemented, for groups that want to be able to designate an "admin" with the authority to remove whoever. For now, while we don't have a permissions system, we're placing all users on equal footing, and so that involves voting.

Reply:
Okay, so groups with 2 members will never be able to exclude each other. The only way will be by removing themselves in group settings with "Leave Group". Roger, I'll let you know when the change is made.

Reply:
Unresolving this not because there are any issues, but because this thread is important for helping us remember a summary of all of the major concerns and issues that we discovered with respect to removing members.

So it's a useful reference to not lose.